File: The Trade Desk/48.rotate-image.java
/**
 * Algorithm:
 * Rotate 90 degrees clockwise by transposing the matrix and reversing each row.
 * This performs the rotation in place.
 *
 * Complexity:
 * Time O(n^2), space O(1).
 */
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;
        for (int r = 0; r < n; r++) {
            for (int c = r + 1; c < n; c++) {
                int tmp = matrix[r][c];
                matrix[r][c] = matrix[c][r];
                matrix[c][r] = tmp;
            }
        }
        for (int[] row : matrix) {
            reverse(row);
        }
    }

    private void reverse(int[] row) {
        int left = 0;
        int right = row.length - 1;
        while (left < right) {
            int tmp = row[left];
            row[left++] = row[right];
            row[right--] = tmp;
        }
    }
}

File: Nvidia/48.rotate-image.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        /*
        Approach:
        A clockwise 90-degree rotation is transpose plus reversing each row.
        Transpose swaps across the main diagonal, then row reversal places the
        columns in the rotated order.

        Complexity: O(n^2) time and O(1) extra space.
        */
        int n = (int)matrix.size();
        for (int r = 0; r < n; ++r) {
            for (int c = r + 1; c < n; ++c) swap(matrix[r][c], matrix[c][r]);
        }
        for (auto& row : matrix) reverse(row.begin(), row.end());
    }
};
